### PR TITLE
feat(Core/Hook): New hook OnBeforePlayerLogout

### PR DESCRIPTION
**Pull request by NathanHandley**
## Changes Proposed:
Adding a new hook "OnBeforePlayerLogout" which fires _before_ the logout logic is executed for a player inside WorldSession::LogoutPlayer.  This is to support a mod I made which allows a character to change their class, which you can find [here](https://github.com/NathanHandley/mod-multi-class).

There is logic to perform before the logout save event, which made utilizing the existing OnPlayerLogout non-viable (visual item changes on class switch were not being pushed and registered on the client).  Since the save hook does not have the "logout" bool available, I was left with a choice to add this new hook or update/overload that hook.  It seems that a OnBeforeLogout could be useful for more hypothetical use cases, so I elected with this approach.  Also, this allows it to work in the situation that save on logout isn't enabled for some reason.

## Issues Addressed:
none

## SOURCE:
n/a

## Tests Performed:
- Tested on both an Ubuntu 22.04 and Windows 10 Pro server

## How to Test the Changes:
The mod [here](https://github.com/NathanHandley/mod-multi-class) uses this hook.  Install that mod, and switch your class while wearing gear.

## Known Issues and TODO List:
none

### CODE REVIEW

File: src/server/game/Scripting/ScriptDefines/PlayerScript.cpp [hunk 1]
```diff
@@ -361,6 +361,14 @@ void ScriptMgr::OnPlayerLoadFromDB(Player* player)
     });
 }
 
+void ScriptMgr::OnPlayerPreLogout(Player* player)
```
Comment (Kitzunu) [inline src/server/game/Scripting/ScriptDefines/PlayerScript.cpp:364]:
```suggestion
void ScriptMgr::OnPlayerBeforeLogout(Player* player)
```

Comment (NathanHandley) [reply src/server/game/Scripting/ScriptDefines/PlayerScript.cpp]: Changed per review

File: src/server/game/Scripting/ScriptDefines/PlayerScript.cpp [hunk 2]
```diff
@@ -361,6 +361,14 @@ void ScriptMgr::OnPlayerLoadFromDB(Player* player)
     });
 }
 
+void ScriptMgr::OnPlayerBeforeLogout(Player* player)
```
Comment (Kitzunu) [inline src/server/game/Scripting/ScriptDefines/PlayerScript.cpp:364]:
```suggestion
void ScriptMgr::OnBeforePlayerLogout(Player* player)
```

My bad! Doing several things at the same time.

Comment (NathanHandley) [reply src/server/game/Scripting/ScriptDefines/PlayerScript.cpp]: No biggie.  It's an easy change to make.  Thanks for reviewing.

**Review by Kitzunu — APPROVED** [review]

**Pull request merged**